### PR TITLE
ASM-2066 Support Post OS Add to domain for Window

### PR DESCRIPTION
**Pull request by sushilrai**
Added new class and template for adding the host to the domain.

### CODE REVIEW

File: manifests/nic/nic_ip_settings.pp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  $exec_lck = "${vardir}/postinstall.lck"
+
+
+# Validate Parameters
+  validate_string($macaddress)
+  validate_string($subnet)
+  validate_string($gateway)
+  validate_string($nic_adapter_name)
+
+  notice($exec_lck)
```
Comment (gavin-scott) [inline manifests/nic/nic_ip_settings.pp]: Is `$exec_lck` being used for anything other than the `notice`?

File: templates/domain_config.ps1.erb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Powershell script template for configuring domain configuration
+
+<% require 'asm/cipher'
+   require 'asm/util' %>
+
+$domain          = '<%= @ipaddress %>',
+$username        = '<%= @username %>',
+$password        = '<%= ASM::Cipher.decrypt_string(@password) %> %>',
```
Comment (gavin-scott) [inline templates/domain_config.ps1.erb]: - Will this leave the template file with the plaintext password on the Windows host? We need to figure out some way of deleting it or other way of passing this through securely.
- What if the password has a `'` in it? AFAICT in powershell `'` is the only special character in single-quoted strings so I think doing `$var.gsub("'", "''")` would be good enough to escape them. We should do this on all the values used with single-quoted strings.

File: templates/domain_config.ps1.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+$machine_ou      = '<%= @machine_ou %>',
+$resetpw         = '<%= @resetpw %>',
+$reboot          = '<%= @reboot %>',
+$join_options    = '<%= @join_options %>',
+
+# Since the powershell command is combersome, we'll construct it here for clarity... well, almost clarity
+#
+(Get-WmiObject -Class Win32_ComputerSystem).JoinDomainOrWorkGroup('${domain}',${password},'${username}@${domain}',${_machine_ou},${join_options})
```
Comment (gavin-scott) [inline templates/domain_config.ps1.erb]: Does doing '${domain}' work? I though single-quoted strings didn't do variable interpolation e.g.:

    PS C:\Users\Gavin_Scott> $domain = 'Americas'
    PS C:\Users\Gavin_Scott> echo '${domain}'

Also, what is `${_machine_ou}`? The variable name on line 9 doesn't have an _, does this work also?

**Comment by aimers1975** [comment]
@gavin-scott Are you good with these changes, or anything else for @sushilrai to address?

File: manifests/domain/domain_config.pp [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# == Class: domain_config
+#
+# Full description of class domain_config here.
```
Comment (gavin-scott) [inline manifests/domain/domain_config.pp]: Please update this.

File: manifests/domain/domain_config.pp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#
+# === Parameters
+#
+# [*domain*]
+#   Explanation of what this parameter affects and what it defaults to.
+#   e.g. "Specify one or more upstream ntp servers as an array."
```
Comment (gavin-scott) [inline manifests/domain/domain_config.pp]: Needs to be updated.

File: templates/domain_config.ps1.erb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+$username        = '<%= @username %>'
+$password        = '<%= ASM::Cipher.decrypt_string(@password) %>'
+
+# Since the powershell command is combersome, we'll construct it here for clarity... well, almost clarity
+#
+
+$domainfqdd = $domainfqdd.gsub("'", "''")
+$domainname = $domainname.gsub("'", "''")
+$username = $username.gsub("'", "''")
+$password = $password.gsub("'", "''")
```
Comment (gavin-scott) [inline templates/domain_config.ps1.erb]: This needs to be done in the ruby code not here. I.e. I was thinking for e.g. line 9:

    $password        = '<%= ASM::Cipher.decrypt_string(@password).gsub("'", "''") %>'

**Comment by gavin-scott** [comment]
I still have open questions on:

1. What is the `notice($exec_lck)` stuff about?
2. How do we prevent plaintext domain passwords sitting on the remote host's filesystem?

**Comment by sushilrai** [comment]
notice($exec_lck) has been removed. This was added just add message in case lock is already added.

There are two options
1. Encrypt the password using powershell capability, but for that we need the access to a windows machine using which encrypted password can be generated. Currently we do not mandate to have a remote Windows machine.

2. Password is added to resouces.txt file in the puppetlabs  state/run directory. We can remove this file at the end of the manifest. 

file { 'c:\programdata\puppetlabs\puppet\var\state\resources.txt':
   ensure => absent
}

We can add this change along with up-coming windows post-installation changes

**Comment by gavin-scott** [comment]
Are you proposing that we merge this as is and handle securing the password subsequently? We can do that but I want to make sure it doesn't get forgotten, we can put in a jira for it. I don't know if @ripienaar might have ideas on securing passwords like this, I can ping him. We've figured out reasonable methods for the puppet apply stuff running on master but I don't think we've dealt with it on remote agents much before.

Also, I don't think domain_config.ps1.erb works, see the comment I had on lines like: `$password = $password.gsub("'", "''")`

**Comment by ripienaar** [comment]
Can't we do the same environment trick here?  If the exec sets the password in some environment var the powershell can just fetch it there.  This probably wont show up in resources.txt then

**Comment by gavin-scott** [comment]
@ripienaar I'm not sure I follow -- I'm not familiar with resources.txt or what it contains. I think the problem is if we try to set an environment variable with something like:

    exec { 'foo':
      environment => ["PASSWORD=bar"],
      command => '/bin/echo $PASSWORD > /tmp/bar'
    }

then the password would be in plaintext in our node_data yaml file. I'm not sure if the "compiled" manifest  file is also stored somewhere on the agent side, in which case that would be another plaintext leak of the password. If the latter is not a problem, then maybe it's possible to have the decryption done for certain parameters in our ENC? That would allow the node_data yaml to contain encrypted passwords...

**Comment by gavin-scott** [comment]
Merging this with the understanding that the plaintext password issue will be addressed in a subsequent commit.

**Pull request merged**